File: 백준/Gold/12865. 평범한 배낭/평범한 배낭.py
import sys

input = sys.stdin.readline

N, K = map(int, input().split())
dp = [[0]*(K+1) for _ in range(N+1)]
for i in range(N):
    W, V = map(int, input().split())
    for j in range(K + 1):
        if j >= W:
            dp[i][j] = max(dp[i-1][j], dp[i-1][j - W] + V) # 이전에 넣었던 짐에서 현재 짐 넣어보기
        else:
            dp[i][j] = dp[i-1][j]

print(dp[N-1][K])

File: 백준/Gold/12865. 평범한 배낭/평범한 배낭.js
const fs = require("fs");
const filePath = process.platform === "linux" ? "/dev/stdin" : "input.txt";
const input = fs.readFileSync(filePath).toString().trim().split("\n");

const [N, K] = input.shift().split(" ").map(Number);
const arr = input.map((line) => line.split(" ").map(Number));

let dp = Array(K + 1).fill(0);

for (let [w, v] of arr) {
  let tmpDP = [...dp];
  for (let i = 0; i <= K - w; i++) {
    tmpDP[i + w] = Math.max(dp[i + w], dp[i] + v);
  }
  dp = tmpDP;
}

console.log(dp[K]);
